### ISSUE TITLE
ruTorrent Update + Load Optimizations

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#17]
ruTorrent should be updated to latest version.

Need to remember plugins are set correctly (our custom ones brought, and broken ones omitted).

We should also optimize a little bit beyond the typical link css etc. from outside server; Namely the initial loading requests a lot of file exist & accessible checks through rtorrent to enable or disable features; We should skip this, we already know these are always available and if not there's a bigger issue. These checks however are extraordinary slow in rtorrent, slows loading time a lot.

**Comment by MagnaCapax** [comment]
**Cross-reference:** This issue is largely subsumed by #177 (rTorrent + ruTorrent version upgrade: 0.9.8 → 0.15.7, v3.10 → v5.2.x). The ruTorrent update and load optimizations described here will be addressed as part of that coupled upgrade. See #177 for the full research and migration plan.

— Sampsa Pellervoinen 🌱

**Comment by MagnaCapax** [comment]
## Investigation Report

**Issue:** #17 — ruTorrent Update + Load Optimizations
**Investigated:** 2026-04-27
**Status:** PARTIAL (superseded — both halves rolled into active upgrade work)

### Summary

This issue is fully superseded by three active issues, all open and recently updated: #177 (parent — coupled rTorrent + ruTorrent upgrade), #303 (ruTorrent v3.10 → v5.2.x), and #364 (rTorrent 0.9.8 → 0.15.7). The 2026-04-05 adversarial review already verdicted BLOCKED on this issue; 22 days later, the supersession remains real and durable, and no v3.10 code change has landed in `etc/skel/www/rutorrent/` since (only an unrelated hddquota PHP 8 fix on a different path). The performance concern is legitimate but its diagnosis in the issue body is incorrect, and the correct fix only makes sense in the v5.2.x context tracked by #303.

### What The Issue Asks For

Two requests, posted 2024-02-02 by the operator:

1. **Update ruTorrent to latest** with custom plugins preserved and broken plugins omitted.
2. **Optimize page-load latency** by skipping the file-existence checks that ruTorrent runs through rTorrent on every page load — the issue argues those binaries are "always available" so the checks are wasted work and are extraordinarily slow through rTorrent.

### Implementation Status

**Part 1 — version update:** NOT implemented in v3.10 skel. Replaced by full upgrade work:
- `#303` deeply documents v3.10 → v5.2.10 migration: 43 → 47 plugins, hddquota preserved, diskspace excluded, `rpc` removed (httprpc covers it), 8 custom themes need rewriting for v5.2's Bootstrap CSS architecture, three v5.2.x upstream bugs flagged (ratio Priority-Critical #2958/2999, getMinFilePerms #3005/2987, filedrop #2966).
- `#177` is the parent coupled-upgrade issue.
- `#364` covers the rTorrent half (0.9.8 → 0.15.7, 30x RPC performance, API v11).

**Part 2 — load optimization:** NOT implemented and the issue's diagnosis does not match the code:

- `getplugins.php` is 528 lines, rebuilt fresh by upstream in v5.2.x with a different class architecture (autoloader, `CachedEcho::send()`, `FileUtil::getPluginConf()`).
- The `$do_diagnostic` flag (default `true` in both `etc/skel/www/rutorrent/conf/config.php:28` and the deployed template `etc/seedbox/config/template.rutorrent.config:21`) gates only:
  - Line 273: one `findRemoteEXE('stat',...)` call.
  - Lines 449–453: three `rtorrent.external.warning` plugin checks.
- The actual hot path is `getplugins.php:403` — `findRemoteEXE` for `rtorrent.external.error` declarations — which fires **unconditionally** regardless of `$do_diagnostic`. Across 15 plugins this deduplicates to 5 distinct rTorrent XMLRPC round-trips per page load (`php`, `sox`, `mediainfo`, `ffmpeg`, `pgrep`).
- Setting `$do_diagnostic = false` would save at most 4 XMLRPC calls; the 5 unconditional ones survive. The "skip the check" framing in the issue body is therefore mechanically incorrect for v3.10.

**Authoritative config source:** `etc/seedbox/config/template.rutorrent.config` is the per-deployment source of truth. `updateRutorrentConfig()` regenerates `conf/config.php` from this template on every upgrade — a direct edit to the skel `conf/config.php` would be overwritten on the next user upgrade.

### What's Missing

- The v5.2.x upgrade itself (tracked in #303 / #177).
- A session-level cache for the binary-existence probes — the *correct* shape of the load-optimization fix. It should be evaluated when #303 lands (the v5.2.x equivalent of `getplugins.php` may have eliminated the pattern upstream; if not, cache-once-per-session is the right shape).

### Mission Alignment

- **Reliability over heroics (#7):** Patching v3.10 vendor code now creates a divergence that has to be re-resolved when v5.2.x lands. The merged path (defer to #303) is the boring/durable one.
- **Investigate before acting (#8):** The 2026-04-05 adversarial review proved the issue body's diagnosis was wrong. Acting on a wrong diagnosis would have introduced complexity without delivering the claimed performance win.
- **Reuse over re-implement:** Carrying the load-optimization concern forward into #303 reuses that issue's existing migration plan and v5.2.x evaluation rather than duplicating analysis here.
- **No concerns** under liberty/privacy, business longevity, or honesty cardinal values for the proposed close-as-superseded action.

### Devil's Advocate

**Argument FOR keeping #17 open and acting now:**
- The performance pain is real for users still on v3.10 today, and #303 has not landed.
- A two-line `$do_diagnostic = false` template change is a trivial intervention that may help a little.

**Rebuttals:**
- The performance gain from `$do_diagnostic = false` is small (4 of 9 distinct XMLRPC calls saved); the dominant cost (5 unconditional calls) is unchanged, so users would not perceive a meaningful change.
- The fix is patching vendor code (per AGENTS.md, vendor trees are read-only without explicit approval) for a path that gets wholesale-replaced by #303.
- Leaving the issue open with a wrong diagnosis attached to the body ("skip these checks, they're always available") risks a future agent picking it up and re-investigating from scratch. Closing as superseded with a pointer to #303 short-circuits that loop.

**Verdict:** Devil's advocate fails — close as superseded.

### Anti-Pattern Hunt

- **REGEX_FOR_SEMANTICS:** PASS — recommendation is a supersession close. No deterministic semantic enforcement proposed.
- **PREMATURE_OPTIMIZATION:** PASS — the underlying issue *is* a premature optimization (optimizing v3.10 code that's about to be replaced); the recommendation explicitly defers optimization to the v5.2.x evaluation in #303 where the cost/benefit is real and the target is durable.
- **FALSE_COMPLETION_CLAIM:** PASS — recommendation is "do not implement here, this work belongs in #303/#177/#364." No completion claim.
- **OPERATOR_HANDOFF_ANTIPATTERN:** PASS — concrete recommendation given (SHOULD-NOT-IMPLEMENT, close as superseded). Not deferring to operator for a decidable question.
- **FABRICATED_AUTHORITY_OR_PATH:** PASS — every cited file path was opened and verified in this run (`getplugins.php`, `conf/config.php`, `template.rutorrent.config`, ADR directory, `pmssUserUpgradeRutorrent`, plus the active state of #177 / #303 / #364 from the pre-injected open-issues list).
- **FAIL_OPEN_SAFETY_BEHAVIOR:** PASS — N/A. No safety-gating behavior in the recommendation.
- **DRY_RUN_THEATER:** PASS — N/A. Read-only investigation by spec.
- **INVESTIGATION_LOOP_THEATER:** PASS with note — this is the second investigation pass on #17 (after 2026-04-05). The recommendation is to terminate the loop by closing as superseded, not to re-investigate again.

### Related Incidents & Data

- **PMSS git history on the rutorrent skel since 2026-04-05:** one commit (6345380a, hddquota PHP 8 arithmetic fix, Refs #304). No work on `getplugins.php` or `template.rutorrent.config`. Supersession remains real.
- **PHP 8.x compat patcher debt:** #303's body documents a recurring pattern — each PHP 8.2 incompatibility found in the frozen v3.10 vendor tree spawns a new runtime patcher (#278 settings.php, #271 filemanager.php, plus rss/action.php and hddquota/action.php still pending). This is the strongest business case for landing #303 and the strongest reason **not** to spend effort on v3.10 changes.
- **No customer-reported tickets cited as load-time pain on ruTorrent in the 2026-04-05 prior review** — the optimization concern is a real-but-undimensioned latency, not a fire.

### Cross-Issue Analysis

From the pre-injected open-issue list, three direct supersession candidates (all open, all recently active):

| Issue | Title | Comments | Relationship to #17 |
|-------|-------|----------|---------------------|
| **#177** | rTorrent + ruTorrent version upgrade: 0.9.8 → 0.15.7, v3.10 → v5.2.x | 29 | **Parent.** Coupled upgrade. Subsumes both halves of #17. |
| **#303** | ruTorrent upgrade: v3.10 to v5.2.x — investigation and plan | 7 | **Direct duplicate of #17 part 1**, with full migration plan. |
| **#364** | rTorrent upgrade: 0.9.8 to 0.15.7 — critical fixes and 30x RPC performance | 39 | rTorrent half. 30x XMLRPC perf claim directly addresses #17 part 2 *root cause* — the slow rTorrent XMLRPC path is the actual bottleneck. |

Adjacent (not duplicates, but related):
- `#316` (Sustained refactoring: extract utilities, eliminate function_exists guards) — `getplugins.php` at 528 lines violates AGENTS.md 75-line split guideline; refactor naturally lives in #316's scope when v5.2.x lands.
- `#17` *part 2* is solved at the rTorrent layer (#364, 30x RPC perf) more cleanly than at the ruTorrent layer; another reason not to patch v3.10's `getplugins.php`.

**Recommendation:** SUPERSEDES. Close #17, with body pointing to #177 (parent), #303 (UI half), #364 (daemon half). Carry the load-optimization observation as a sub-bullet on #303.

### Dependencies & Blockers

None to act on for #17 itself. Closing it as superseded has no dependencies. The downstream upgrade work (#177/#303/#364) has its own dependency chain (PHP 8.4 compat #361, Debian 13 validation #111, /tmp tmpfs #362) which is properly tracked there.

### Reuse, Refactor & Test Strategy

- **Existing reuse points** — `pmssUserUpgradeRutorrent()` and `updateRutorrentConfig()` already implement the rename-and-replace upgrade pattern correctly. v5.2.x should reuse this scaffolding (#303 documents the exact diffs needed). For #17 specifically, no reuse point applies because the recommendation is "do not implement."
- **Refactor opportunities** — when #303 lands, `getplugins.php`'s four mixed responsibilities (plugin metadata loading, XMLRPC probing, JS assembly, output caching) are a natural decomposition target; the temp-file rendezvous in `findRemoteEXE` / `testRemoteRequests` should be replaced with cached results.
- **Test categories proposed** — N/A for #17. For #303 (where this work actually goes): hermetic unit (config-template regeneration), integration (rolling upgrade on a non-production server), regression (verify the three known v5.2.x upstream bugs are caught: ratio, getMinFilePerms, filedrop), boundary (8 custom themes against v5.2 Bootstrap CSS), failure injection (rTorrent down during `findRemoteEXE`).
- **Minimum-code-for-correctness** — for #17 itself, zero. For the carried-forward load-optimization concern: a session-level cache keyed by binary name with a single-page-load TTL.
- **Mission/business success signal** — Closing #17 reduces the open-issue pool by one stale-but-active item, making operator triage cleaner (cardinal value #9 — continuous improvement). Business signal: avoids spending engineering time on v3.10 patches that get thrown away when #303 ships.

### Architectural Minimization

**Step A — dumbest requirement:** "We already know these are always available, skip the check." Source: issue body, 2024-02-02. Challenge: `php` and `pgrep` are always available on PMSS, but `sox`, `mediainfo`, and `ffmpeg` are optional system packages whose absence should still disable the corresponding plugins (`spectrogram`, `mediainfo`, `screenshots`). "Always available" is wrong in three of five cases. The underlying intent — "stop paying for these checks per page-load" — is valid; the proposed mechanism (skip) is not.

**Step B — DELETE / MERGE / REFACTOR / NULL alternatives:**

- **DELETE candidate 1:** Remove `spectrogram`, `mediainfo`, `screenshots` plugins from the default PMSS bundle. Eliminates 3 of 5 unconditional XMLRPC calls. Status: legitimate option, but already proposed under #303 plugin-pruning evaluation. Not a v3.10 change.
- **DELETE candidate 2:** Remove the `$do_diagnostic` block entirely (set to `false` in template). Status: insufficient — saves only 4 of 9 calls, leaves dominant cost intact. Rejected.
- **MERGE candidate 1:** Fold #17 into #303 as a sub-bullet under the v5.2.x evaluation. **PREFERRED.** This is the recommendation.
- **MERGE candidate 2:** Fold #17 into #364 (rTorrent upgrade). Status: partially justified — the 30x XMLRPC perf in 0.15.7 reduces the per-call cost by an order of magnitude, addressing the "extraordinarily slow in rtorrent" complaint at its root. Reasonable secondary cross-link.
- **REFACTOR candidate 1:** Replace `findRemoteEXE` / `testRemoteRequests` temp-file rendezvous with cached results. Status: belongs in #303's v5.2.x rewrite, not v3.10. Rejected for v3.10.
- **REFACTOR candidate 2:** Extract `getplugins.php` (528 lines) into named modules per AGENTS.md split guideline. Status: belongs in #316 / #303 post-upgrade. Rejected for v3.10.
- **NULL alternative:** What breaks if we just don't act? Users on v3.10 keep paying ~5 XMLRPC round-trips per page load until #303 ships. No data loss, no service outage, no customer-blocking failure. Latency-only.

**Best path:** MERGE candidate 1 (close #17, pointer to #303 with load-optimization note carried forward). DELETE candidate 1 should be evaluated as part of #303 plugin-pruning, not here.

**Step C — minimal design proposal:** Alternative wins. No minimal v3.10 design proposed because the MERGE-into-#303 path delivers the same outcome with no thrown-away work.

### ADR Adherence

`NO_RELEVANT_ADRS:` searched the project ADR directory; 14 ADRs listed (template, cgroup-v2 dual-path detection, shell command guardrails, trust boundaries for internal tools, update step2 user loop and observability, install bootstrap interactivity, reject deb822 apt sources migration, nginx lightweight reverse proxy, evaluate fail2ban, defer default MAC enforcement, defer default unbound DNS caching, defer default ext4 native quota migration, update step2 step error classification policy, defer browser console until packaged and reviewed). None touch ruTorrent, rTorrent, the user-facing UI, plugin loading, or XMLRPC behavior. The recommendation does not propose an architectural decision that would merit a new ADR — closing as superseded is operational hygiene, not architecture.

### Deep Research Gate

`SKIPPED: prior-investigation-count` — issue has been investigated twice already (2026-02-01 cross-reference, 2026-04-05 full adversarial review). Citation: prior 2026-04-05 review plus a comprehensive 2026-03-24 internal investigation on the v5.2.x upgrade path covers every relevant question. Further deep research would be investigation-loop theater on a stale issue whose recommendation is "close as superseded."

### Final AQFM Sweep

Question asked (verbatim):

> Issue #17 (ruTorrent Update + Load Optimizations, opened 2024-02-02): Draft conclusion: Issue is fully superseded by #303 (ruTorrent v3.10->v5.2.x), #364 (rTorrent 0.9.8->0.15.7), and parent #177. Prior 2026-04-05 adversarial review verdict: BLOCKED. Key evidence: etc/skel/www/rutorrent/php/getplugins.php line 403 (unconditional findRemoteEXE), template.rutorrent.config (deployed config source). Draft recommendation: SHOULD-NOT-IMPLEMENT — close as superseded duplicate. What contradicting evidence, missed precedent, or operator decision exists in memory that could revise this draft?

Sources returned (synthesis layer hit a permission wall but the retrieval layer returned source files): communications doctrine, GH issue privacy doctrine, project root instructions, accuracy/anti-hallucination doctrine, opportunistic-loading map, memory-search doctrine, git workflow doctrine, WHMCS billing doctrine. None contradict the draft. Targeted broad search for prior `#17` decisions (`rg -l "17.*supersed|gh-issue-17|issue-17"`) returned only the existing 2026-04-05 lesson and unrelated issue files.

Verdict: **HELD** — no revision needed.

### Adversarial Self-Challenge

**Strongest finding 1:** Issue is fully superseded by #303 + #364 (and parent #177); all three open and active.

- Disproving search 1: `pre-injected open-issue list` (cross-issue analysis section) — checked all three. **Result:** all three issues open, recently updated (29 / 7 / 39 comments respectively). No counter-evidence.
- Disproving search 2 (broader): broad keyword sweep across local research notes for prior decisions on issue #17 — **Result:** only the existing 2026-04-05 lesson hit; no operator directive overriding supersession; no later decision to keep #17 separate.
- Outcome: **held**.

**Strongest finding 2:** The issue's own diagnosis — "skip the file-existence checks, those binaries are always available" — is mechanically incorrect for v3.10 because `findRemoteEXE` for `rtorrent.external.error` runs unconditionally at `getplugins.php:403`.

- Disproving search 1: `grep -n "findRemoteEXE\|testRemoteRequests\|do_diagnostic\|external.error\|external.warning" etc/skel/www/rutorrent/php/getplugins.php` — **Result:** confirmed line 403 sits outside any `$do_diagnostic` guard; lines 273 and 449–453 are inside one. The 2026-04-05 review's reading is correct.
- Disproving search 2 (broader): `git log --oneline -10 etc/skel/www/rutorrent/php/getplugins.php` — **Result:** four commits since import, all permissions/ownership; no functional change since the prior review. No revision needed.
- Outcome: **held**.

**Least-evidenced finding** (selected by lowest citation count): "30x XMLRPC perf in rTorrent 0.15.7 (#364) addresses the 'extraordinarily slow in rtorrent' complaint at its root" — cited only via the issue list summary line, not from a verified upstream benchmark in this run.

- Disproving search 1: broad keyword sweep across local research notes for rTorrent 0.15.7 XMLRPC performance — **Result:** corroborating research note exists from 2026-03-24 ("rtorrent 0.15.7 vs 0.9.8 real-world seedbox improvements") which would back the claim, but I did not open it in this run. The "30x" magnitude originates from #364's title (operator-authored), not from independent verification here.
- Disproving search 2 (broader): no contradicting evidence found in local research notes for the rTorrent 0.15.7 performance claim.
- Outcome: **revised** — soften from "addresses at its root" to "is the structural fix the issue's *part 2* concern actually points to (rTorrent's slow XMLRPC), per #364's premise; verification of the exact magnitude is #364's investigation, not #17's." This change is reflected in the Cross-Issue Analysis section above.

### Recommendation

**SHOULD-NOT-IMPLEMENT** + close as superseded duplicate of #177 (parent), #303 (UI half), #364 (daemon half).

**Reasoning:**

1. The 2024-02-02 issue is mechanically subsumed by the three open upgrade issues. Both halves of #17 (version update, load optimization) belong in those workstreams.
2. The 2026-04-05 adversarial review reached the same verdict (BLOCKED) and 22 days later nothing has changed in the relevant code path.
3. The issue's own diagnosis ("skip these checks, they're always available") is mechanically wrong against the v3.10 source — the slow XMLRPC calls run unconditionally at `getplugins.php:403`, outside the `$do_diagnostic` flag the issue implicitly targets. Leaving the wrong diagnosis attached to an open issue invites future agents to re-investigate from scratch.
4. The legitimate concern — per-page-load XMLRPC binary probing — has the correct fix shape (session-level cache, or upstream removal in v5.2.x) tracked under #303, plus the underlying rTorrent XMLRPC slowness addressed at the daemon layer in #364.
5. AGENTS.md treats vendor trees as read-only; patching v3.10 vendor code requires explicit approval, and any such patch is discarded when #303 ships.

**Suggested closing comment for the operator:** "Closing as superseded by #177 (parent) / #303 (UI) / #364 (daemon). Load-optimization concern carried forward as a sub-bullet on #303 — when v5.2.x's plugin loader is reviewed, evaluate whether the equivalent of `findRemoteEXE` / `testRemoteRequests` still fires per page load; if so, replace with session-level cache (not skip) since not all probed binaries are guaranteed present."

### Additive Recommendation Flags

- `RECOMMEND_NEW_ISSUE` — **decline.** The load-optimization concern fits cleanly as a sub-bullet on #303 rather than warranting a fresh issue. A new issue would just be #17 reopened under a new number.
- `RECOMMEND_ADR` — **decline.** No architectural decision rises to ADR threshold here. The ADR-worthy decisions (vendor tree read-only policy, upgrade backup naming convention, template-vs-skel config ownership) are already implicit in AGENTS.md and prior #303 documentation.

**Comment by MagnaCapax** [comment]
## Closing as Superseded

Closing per the [2026-04-27 investigation report](https://github.com/MagnaCapax/PMSS/issues/17#issuecomment-4324425584) and the [2026-04-05 adversarial review](https://github.com/MagnaCapax/PMSS/issues/17#issuecomment-4189445464) — both halves of this issue are subsumed by active workstreams.

**Successors:**
- [#177](https://github.com/MagnaCapax/PMSS/issues/177) — parent: coupled rTorrent + ruTorrent upgrade.
- [#303](https://github.com/MagnaCapax/PMSS/issues/303) — UI half: ruTorrent v3.10 → v5.2.x with full migration plan.
- [#364](https://github.com/MagnaCapax/PMSS/issues/364) — daemon half: rTorrent 0.9.8 → 0.15.7 (addresses the underlying slow-XMLRPC root cause).

**Load-optimization concern carried forward** as a sub-bullet on #303: when v5.2.x's plugin loader is reviewed, evaluate whether the equivalent of `findRemoteEXE` / `testRemoteRequests` still fires per page load; if so, replace with session-level cache (not skip) — not all probed binaries (`sox`, `mediainfo`, `ffmpeg`) are guaranteed present, so blanket-skipping is incorrect.

The v3.10 `getplugins.php:403` patch path is a vendor-tree change for code that gets wholesale-replaced when #303 lands — no durable value.

— Väinämöinen
